source/type/vehicle_journey.cpp

```diff
@@ -267,31 +278,49 @@ std::set<StopPoint*> VehicleJourney::get_sections_stop_points(const StopArea* st
      *  - s1/s2 from the 5th to the 10th,
      *  - s1/s2/s5 from the 11th to the 15th
      * */
-    std::set<StopPoint*> res;
+    std::set<uint16_t> res;
     boost::optional<uint16_t> start_idx;
+    boost::optional<uint16_t> last_area_idx;
+    bool section_ending = false;
     const auto* base_vj = this->get_corresponding_base();
     const auto* vj = base_vj ? base_vj : this;
     for (const auto& st : vj->stop_time_list) {
         // The line section is using stop_areas so we make sure we have one
         if (!st.stop_point || !st.stop_point->stop_area) {
             continue;
         }
+        const auto* sa = st.stop_point->stop_area;
+
+        // Must close section before potentially starting a new one
+        if (section_ending && *last_area_idx != sa->idx) {
+            for (uint16_t i = *start_idx, last = st.order() - 1; i <= last; ++i) {
+                res.insert(i);
+            }
+            // the section is finished, we are no more in a section
+            start_idx = boost::none;
+            section_ending = false;
+        }
 
         // we want the shortest section, thus, we set the idx each
-        // time we see it
-        if (st.stop_point->stop_area->idx == start_stop->idx) {
+        // time we see it, except if we are in the same area than the start idx
+        // Reset start_idx if the previous area was different
+        if (sa->idx == start_stop->idx && (!start_idx || !last_area_idx || *last_area_idx != sa->idx)) {
```

Comment:
I would find it easier to read with a `section_starting` flag with pretty much the same logic than section_ending.

Reply:
I hope it's better like this